homeassistant/components/squeezebox/media_player.py

```diff
@@ -126,18 +140,75 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
     lms = Server(async_get_clientsession(hass), host, port, username, password)
     known_servers.add(ipaddr)
 
-    players = await lms.async_get_players()
-    if players is None:
-        raise PlatformNotReady
-    media_players = []
-    for player in players:
-        media_players.append(SqueezeBoxDevice(player))
+    async def _discovery(now=None):
+        """Discover squeezebox players by polling server."""
+
+        async def _discovered_player(player):
+            """Handle a (re)discovered player."""
+            entity = next(
+                (
+                    known
+                    for known in known_players
+                    if known.unique_id == player.player_id
+                ),
+                None,
+            )
+            if entity and not entity.available:
+                # check if previously unavailable player has connected
+                await player.async_update()
+                entity.available = player.connected
+            if not entity:
+                _LOGGER.debug("Adding new entity: %s", player)
+                entity = SqueezeBoxEntity(player)
+                known_players.append(entity)
+                async_add_entities([entity])
+
+        players = await lms.async_get_players()
+        if players:
+            for player in players:
+                hass.async_create_task(_discovered_player(player))
+
+        hass.helpers.event.async_call_later(DISCOVERY_INTERVAL, _discovery)
+
+    _LOGGER.debug("Adding player discovery job for LMS server: %s", ipaddr)
+    hass.async_add_job(_discovery)
+
+    @callback
+    async def start_server_discovery():
+        """Start a server discovery task."""
+
+        @callback
```

Comment:
Since this is `async` it seem like it doesn't need to be decorated with `@callback`?